### PR TITLE
Catalyst/PennyLane adjoint unification

### PR DESCRIPTION
**Pull request by dime10**
Restructures adjoint to inherit from PL adjoint for the single op case. The multi op case is not 1:1 compatible because PL only ever instantiates the Adjoint class on a single base operation. This means that it returns a _list_ of `Adjoint` instances while Catalyst intends to return a single `Adjoint` instance with a list of ops nested inside of it.

Additionally:
- Rename Adjoint class to HybridAdjoint
- Allow adjoint to be used outside of qjit/qnode similar to PL allowing operations to be instantiated anywhere
- Refactor adjoint nested wrappers into a callable
- Add eager (`lazy=False`) adjoint support for single ops
- Include the PL test suite for adjoint in Catalyst to test the better integrated behaviour

[sc-60922]
[sc-60923]

### CODE REVIEW

File: frontend/test/pytest/test_adjoint.py [hunk 1]
```diff
@@ -17,566 +17,1652 @@
...
-            qml.RX(theta / ub, wires=0)
 
-            @while_loop(lambda counter: counter < ub)
-            def while_loop_inner(counter):
-                qml.RZ(ub / 5, wires=0)
-                return counter + 1
+@pytest.fixture(scope="function")
+def use_legacy_opmath():
+    with qml.operation.disable_new_opmath_cm() as cm:
+        yield cm
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/test/pytest/test_adjoint.py:579]: I think we've been supporting it for a while, PL still tests both though hence this fixture.

**Review by erick-xanadu — APPROVED** [review]
I'm going to approve, but if you do add a decomposition, can you ping me again for review? Thanks!

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 2]
```diff
@@ -440,6 +511,25 @@ def wires(self):
...
 
 
+# TODO: Should we inherit from AdjointOperation? Maybe better to mixin the different behaviours like
+# PennyLane does (AdjointOperation, AdjointObs, AdjointOperationObs).
+class Adjoint(AdjointOperation, HybridAdjoint):
+    """This class provides near identical behaviour as PennyLane for adjoint instances with only a
+    single base operation. Additionally, it provides the same functionality as HybridAdjoint."""
+
+    def __init__(self, base_op, tracing_artifacts):
+        AdjointOperation.__init__(self, base_op)
```
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:521]: Can we use the `Adjoint.decomposition` method now?

File: frontend/test/pytest/test_adjoint.py [hunk 3]
```diff
@@ -17,566 +17,1652 @@
...
+            return qml.expval(qml.PauliY(0))
 
-    expected = circuit()
-    observed = qjit(circuit)()
-    assert_allclose(expected, observed)
+        x = jax.numpy.array(0.234)
+        expected_res = jax.numpy.sin(x)
+        expected_grad = jax.numpy.cos(x)
+        assert qml.math.allclose(circ(x), expected_res)
+        assert qml.math.allclose(jax.grad(circ)(x), expected_grad)
```
Comment (maliasadi) [inline frontend/test/pytest/test_adjoint.py:1667]: A thorough testing of adjoint :heart:

**Review by maliasadi — COMMENTED** [review]
Thanks @dime10! Nice work :100:

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 4]
```diff
@@ -440,6 +511,25 @@ def wires(self):
         return total_wires
 
 
+# TODO: Should we inherit from AdjointOperation? Maybe better to mixin the different behaviours like
+# PennyLane does (AdjointOperation, AdjointObs, AdjointOperationObs).
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:515]: As far as I know we've had no support for observables so far

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 5]
```diff
@@ -440,6 +511,25 @@ def wires(self):
...
+        AdjointOperation.__init__(self, base_op)
+        HybridAdjoint.__init__(self, *tracing_artifacts)
+
+    def _flatten(self):
+        tracing_artifacts = (self.in_classical_tracers, self.out_classical_tracers, self.regions)
+        return (self.base, tracing_artifacts), tuple()
+
+    @classmethod
+    def _unflatten(cls, data, _):
+        return cls(*data)
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:530]: This class has more meta-data than its base class, so the PyTree methods need to be overridden

File: frontend/test/pytest/test_adjoint.py [hunk 6]
```diff
@@ -17,566 +17,1652 @@
...
-                    [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j, 0.99500417 - 0.09983342j],
-                ]
-            ),
-            wires=[0, 1],
-        )
+from catalyst import adjoint, cond, for_loop, qjit, while_loop
 
-    verify_catalyst_adjoint_against_pennylane(func, qml.device(backend, wires=2))
 
+class CatalystTests:
```
Comment (maliasadi) [inline frontend/test/pytest/test_adjoint.py:30]: I cannot test this class locally without the following change,
```suggestion
class TestCatalyst:
```

Comment (dime10) [reply frontend/test/pytest/test_adjoint.py:30]: Good catch, thanks!

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 7]
```diff
@@ -403,12 +368,118 @@ def trace_quantum(self, ctx, device, trace, qrp) -> QRegPromise:
...
+            self.single_op = True
+        elif isinstance(target, Callable):
+            # Case 3: User passed an arbitrary callable that will instantiate operations.
+            # We want to create a callable that will generate an "opaque" Adjoint object.
+            # This object differs from the Adjoint in PennyLane because that one can only be
+            # instantiated on single operations.
+            self.callee = target
+            self.single_op = False
+        else:
+            raise ValueError(f"Expected a callable or a qml.Operator, not {target}")
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:399]: Of course, the user could supply `catalyst.adjoint` with an arbitrary input

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 2 above)
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:521]: For single ops I believe so

Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:521]: I tested this locally and we now use `Adjoint.decomposition` with single ops :partying_face:

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 8]
```diff
@@ -49,6 +50,8 @@
 )
 from catalyst.tracing.contexts import EvaluationContext
 
+pl_adjoint_module = sys.modules["pennylane.ops.op_math.adjoint"]
```
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:53]: nice hack! :)

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 9]
```diff
@@ -439,6 +521,64 @@ def wires(self):
         total_wires = sum((op.wires for op in self.regions[0].quantum_tape.operations), [])
         return total_wires
 
+    def decomposition(self):
+        """Resolve the Adjoint region by propagating the adjoint modifier to nested operations
+        in reverse."""
+        assert len(self.regions) == 1, "Expected a single nested region for HybridAdjoint"
+
+        # While catalyst.adjoint would be just as valid since it is PL compatible for single ops,
+        # going to PennyLane's adjoint skips unnecessarily re-queing the base operation in each
```
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:530]:
```suggestion
        # going to PennyLane's adjoint skips unnecessarily re-queuing the base operation in each
```

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 10]
```diff
@@ -439,6 +521,64 @@ def wires(self):
...
+        # HybridOp's nested tape when qjit is not active.
```
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:531]: Thanks for the comment! :+1:

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 11]
```diff
@@ -439,6 +521,64 @@ def wires(self):
...
+    """Replicate mixin class structure from PennyLane for Observables."""
+
+
+class AdjointOpObs(Adjoint, pl_adjoint_module.AdjointOpObs, HybridAdjoint):
+    """Replicate mixin class structure from PennyLane for Operations that are also Observables."""
+
+
+class AdjointBase(Adjoint, pl_adjoint_module.Adjoint, HybridAdjoint):
+    """Replicate mixin class structure from PennyLane for an unkown Operator type."""
+
```
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:581]: :tada:

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 12]
```diff
@@ -439,6 +521,64 @@ def wires(self):
...
+        return cls(*data)
+
+
+# HybridAdjoint is also mixed in because the PL class needs to sit between Adjoint & HybridAdjoint.
+# Thus Adjoint cannot be made to inherit from HybridAdjoint directly.
+class AdjointOperation(Adjoint, pl_adjoint_module.AdjointOperation, HybridAdjoint):
+    """Replicate mixin class structure from PennyLane for Operations."""
+
+
+class AdjointObs(Adjoint, pl_adjoint_module.AdjointObs, HybridAdjoint):
```
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:571]: I may be missed it in `test_adjoint.py`; where is this adjoint(obs) tested?

**Review by maliasadi — APPROVED** [review]
Nice work! :partying_face:

File: frontend/catalyst/api_extensions/__init__.py [hunk 13]
```diff
@@ -35,7 +35,7 @@
 from catalyst.api_extensions.error_mitigation import mitigate_with_zne
 from catalyst.api_extensions.function_maps import vmap
 from catalyst.api_extensions.quantum_operators import (
-    Adjoint,
+    HybridAdjoint,
```
Comment (maliasadi) [inline frontend/catalyst/api_extensions/__init__.py:38]: Just a reminder to add this PR to the changelog before merging ;)

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 12 above)
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:571]: Good question, I don't know. I would hope it's part of the PL test suite that I copied over 👀

File: frontend/catalyst/api_extensions/__init__.py (hunk 13 above)
Comment (dime10) [reply frontend/catalyst/api_extensions/__init__.py:38]: Good point!

File: frontend/catalyst/device/decomposition.py [hunk 14]
```diff
@@ -35,6 +36,7 @@
     rotations_and_diagonal_measurements,
 )
 
+from catalyst.api_extensions import HybridAdjoint
 from catalyst.api_extensions.quantum_operators import QCtrl
```
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/device/decomposition.py:40]: `api_extensions` as a submodule provides its own public methods and objects to be used by the rest of the catalyst package. In such a scheme modules outside of the `api_extensions` folder (not counting tests) should not need to use objects not available directly from the `api_extesions`.

File: frontend/catalyst/api_extensions/__init__.py (hunk 13 above)
Comment (dime10) [reply frontend/catalyst/api_extensions/__init__.py:38]: Done 👍

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 9 above)
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py]: Done

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 12 above)
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:571]: It seems to be tested in a few places such as [here](https://github.com/PennyLaneAI/catalyst/blob/fc1b0320ed347fa9f69c68e506c7155f87db2899/frontend/test/pytest/test_adjoint.py#L1201). There are also tests for HermitianObs :+1:

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 15]
```diff
@@ -228,54 +236,13 @@ def loop_fn(i):
...
-        )
-
-    if isinstance(f, Callable):
+    adj = AdjointCallable(f, lazy=lazy)
 
-        def _callable(*args, **kwargs):
-            return _call_handler(*args, _callee=f, **kwargs)
+    if isinstance(f, Operator):
+        # Return an instantiated version of the Adjoint class if we receive an operator instance.
+        adj = adj()
```
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:243]: minor:
```suggestion
    # Return an instantiated version of the Adjoint class if we receive an operator instance.
    return adj() if isinstance(f, Operator) else adj
```

**Pull request merged**